### PR TITLE
Add module to support mirtop summary output

### PR DESCRIPTION
**Pull request by oneillkza**
## If this PR is for a new module

 - [x] There is example tool output for tools in the https://github.com/ewels/MultiQC_TestData repository
 - [x] Code is tested and works locally (including with `--lint` flag)
 - [x] `CHANGELOG.md` is updated
 - [x] `README.md` is updated
 - [x] `docs/README.md` is updated with link to below
 - [x] `docs/modulename.md` is created
 - [x] Everything that can be represented with a plot instead of a table is a plot
 - [x] Report sections have a description and help text (with `self.add_section`)
 - [x] There aren't any huge tables with > 6 columns (explain reasoning if so)
 - [ ] Each table column has a different colour scale to its neighbour, which relates to the data (eg. if high numbers are bad, they're red); no, but columns with the same scale are the same colour, since they are correlated with each other
 - [x] Module does not do any significant computational work

### CODE REVIEW

**Comment by oneillkza** [comment]
Also note that this has a side-dependency on the data added in ewel/MultiQC_TestData#92

File: docs/modules/mirtop.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+---
+Name: mirtop 
+URL: https://github.com/miRTop
+Description: >
```
Comment (ewels) [inline docs/modules/mirtop.md]: Whitespace here could do weird things. Better instead to simply have:

```yaml
Description: A command line tool to annotate miRNAs and isomiRs using a standard naming
```

File: docs/modules/mirtop.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+---
+Name: mirtop
```
Comment (ewels) [inline docs/modules/mirtop.md]: With capitalisation? eg. `miRTop`? That's ok here - this is used to generate the MultiQC website.

File: multiqc/modules/mirtop/mirtop.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+# Initialise the logger
+log = logging.getLogger(__name__)
+
+class MultiqcModule(BaseMultiqcModule):
+
+    def __init__(self):
+
+        # Initialise the parent object
+        super(MultiqcModule, self).__init__(name='mirtop',
+        anchor='mirtop', target='Subread mirtop',
```
Comment (ewels) [inline multiqc/modules/mirtop/mirtop.py]: I guess that `Subread mirtop` is a copy and paste error and should just be `miRTop`. `name` can also use the funky caps.

File: multiqc/modules/mirtop/mirtop.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        self.mirtop_data = self.ignore_samples(self.mirtop_data)
+
+        if len(self.mirtop_data) == 0:
+            raise UserWarning
+
+        log.info("Found {} reports".format(len(self.mirtop_data)))
+
+        # Write parsed report data to a file
+        self.write_data_file(self.mirtop_data, 'multiqc_mirtop')
+        #log.info(self.mirtop_data)
```
Comment (ewels) [inline multiqc/modules/mirtop/mirtop.py]: Please remove commented out code 👍

File: multiqc/modules/mirtop/mirtop.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+        # Create very basic summary table
+        self.mirtop_stats_table()
+
+        # Create comprehensive beeswarm plots of all stats 
+        self.add_section (
+             name = 'IsomiR Summary Statistics',
+             anchor = 'mirtop-stats',
+             description = "This module parses the summary data generated by <code>mirtop</code>. ",
```
Comment (ewels) [inline multiqc/modules/mirtop/mirtop.py]: The description is parsed as markdown now, so you can just use backticks instead of HTML `<code>`.

File: multiqc/modules/mirtop/mirtop.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        
+        file_names = list()
+        parsed_data = dict()
+        for l in f['f'].splitlines():
+            s = l.split(",")
+            if len(s) < 2 or s[1]=="category":
+                continue
+            else:
+                if len(s) > 0:
+                    #log.info(s)
```
Comment (ewels) [inline multiqc/modules/mirtop/mirtop.py]: Please remove 👍

File: multiqc/modules/mirtop/mirtop.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+                    parsed_data[s[1]] = float(s[3])
+                    s_name = s[2]
```
Comment (ewels) [inline multiqc/modules/mirtop/mirtop.py]: Please pass custom sample names through `self.clean_s_name()` so that features like `-d` and custom sample name cleaning work. See http://multiqc.info/docs/#custom-sample-names

File: multiqc/modules/mirtop/mirtop.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        headers['isomiR_sum'] = {
+            'title': 'IsomiR reads',
+            'description': 'read count summed over all isomiRs in sample',
+            'scale': 'PuBu',
+        }
+        headers['ref_miRNA_sum'] = {
+            'title': 'Reference reads',
+            'description': 'read count summed over all reads mapping to the reference form of a miRNA',
+            'scale': 'PuBu'#,
+            #'shared_key': 'read_count'
```
Comment (ewels) [inline multiqc/modules/mirtop/mirtop.py]: I was going to suggest that you add this - any reason it's commented out? It should be used for all columns that specify read counts, so that their background bars share the same scale.

Please also use the `read_count_modifier` syntax for these columns, eg:
```python
'title': '{} Reads'.format(config.read_count_prefix),
'description': 'Number of reads ({})'.format(config.read_count_desc),
'modify': lambda x: x * config.read_count_multiplier,
```

See http://multiqc.info/docs/#step-3-adding-to-the-general-statistics-table

File: multiqc/modules/mirtop/mirtop.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+        }
+        headers['read_count'] = {
+            'title': 'Total reads',
+            'description': 'all aligned reads',
+            'scale': 'PuBu'#,
```
Comment (ewels) [inline multiqc/modules/mirtop/mirtop.py]: Please use different colour scales for adjacent columns. This makes it easier to skim read the report quickly once that you're familiar with a type of report.

File: multiqc/modules/mirtop/mirtop.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+            'title': 'IsomiR %',
+            'description': 'percentage of reads mapping to non-canonical forms of a microRNA',
+            'min':0,
+            'max':100,
+            'suffix':'%',
+            'scale': 'RdYlGn'#,
+            #'shared_key': 'read_count'
+        }
+
+
```
Comment (ewels) [inline multiqc/modules/mirtop/mirtop.py]: 1rs9jw

😆

File: multiqc/utils/config_defaults.yaml [hunk 11]
```diff
@@ -468,3 +468,7 @@ module_order:
             - RNA
             - DNA
             - methylation
+    - mirtop:
+        module_tag:
+            - RNA
```
Comment (ewels) [inline multiqc/utils/config_defaults.yaml]: Feel free to make new tags :) I think `mirna` and `smrna` would be better here as it's quite a specific tool.

File: multiqc/utils/search_patterns.yaml [hunk 12]
```diff
@@ -232,6 +232,8 @@ methylQA:
 disambiguate:
     contents: 'unique species A pairs'
     num_lines: 2
+mirtop:
+    contents: 'mirtop'
```
Comment (ewels) [inline multiqc/utils/search_patterns.yaml]: Is this just captured from standard out? This needs `shared: true` if so, as stdout log files are sometimes concatenated for some people.

Summary result files in JSON or YAML with consistent filenames are the best 😉

File: docs/modules/mirtop.md [hunk 13]
```diff
@@ -0,0 +1,13 @@
+---
+Name: mirtop 
+URL: https://github.com/miRTop
```
Comment (ewels) [inline docs/modules/mirtop.md]: The tool URL is https://github.com/miRTop/mirtop really.

File: multiqc/modules/mirtop/mirtop.py [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+log = logging.getLogger(__name__)
+
+class MultiqcModule(BaseMultiqcModule):
+
+    def __init__(self):
+
+        # Initialise the parent object
+        super(MultiqcModule, self).__init__(name='mirtop',
+        anchor='mirtop', target='Subread mirtop',
+        href='https://github.com/miRTop',
```
Comment (ewels) [inline multiqc/modules/mirtop/mirtop.py]: Should be https://github.com/miRTop/mirtop

File: multiqc/utils/search_patterns.yaml (hunk 12 above)
Comment (oneillkza) [reply multiqc/utils/search_patterns.yaml]: @lpantano could best answer about how standard the file name will be / what would be the best patterns to match one. I did get the impression that less ambiguity would be better!

**Comment by oneillkza** [comment]
Sounds good -- most of these changes look fairly minor. I've been a little busy over the past few weeks (and am going travelling shortly), but will try to get to them soon!

I can for sure chunk up the plots. The reason for beeswarm plots is that these are for comparison across libraries (of which there may be dozens to hundreds). Stacked barplots might work for some of the proportion values, but might also get pretty unwieldy considering there would need to be one bar for every library. I'll give it a think and see what works.

File: multiqc/utils/search_patterns.yaml (hunk 12 above)
Comment (lpantano) [reply multiqc/utils/search_patterns.yaml]: Hi,

thabks for the comment. This tool will put the name mirtop on the top of the file, so it is not stdout.

said that, can you point me to an example of a tool with json in multiqc to try to adapt our tool output to that?

Thanks!

**Comment by oneillkza** [comment]
Actually, re picking the best visualization; @lpantano, would you mind generating an example file with a few samples in (maybe three or so). Most of the visualisations really only make sense with multiple samples.

If you have a real-world example I can play with (TCGA results?) that would be great, too.

**Comment by oneillkza** [comment]
@ewels, I believe I've addressed all your in-line comments. The two things remaining are the search patterns and deciding how best to split out the plots. I think these need input from you and Lorena.

**Comment by lpantano** [comment]
@oneillkza thanks for doing the updates! One think that can be changed is to try to split the figure into 3, one for:

* *_sum
* *_counts
* *_mean

and use a barplot for each sample. Is the only I can think of,

I'll prepare an example with 4 samples tomorrow or if you have already one, go ahead :P

@ewels, please, let me know if you have any example for a different kind of output for JSON or YAML already integrated in multiqc. That would help a lot to move forward.

Cheers

**Comment by ewels** [comment]
Sure, there are a few, but Salmon is the one that springs to mind (see the [search patterns](https://github.com/ewels/MultiQC/blob/master/multiqc/utils/search_patterns.yaml) to find others). But if you’re using JSON or YAML then the parsing is basically a single line of code in Python. That’s the beauty of it :)

For general log advice, I wrote a blog post with my thoughts ages ago, which could be useful: http://tallphil.co.uk/writing-good-log-files/

Shout if I can be of any help!

Phil

**Comment by ewels** [comment]
NB: Specifically, things that are often missed off but are good to have within the log file as keys are:

- The tool name and version
- The command used
- Any parameters used
- The input file names (as part of the above presumably)

**Comment by lpantano** [comment]
@ewels thanks for this!

Maybe I need to change the whole idea. This output is coming from a subcommands `stats` like `samtools stats`. So it is not technical log. 

What I can do is to change that to have a real log file that is generated when the main function is running and that will end up being the input of multiqc. I can modify to be a log in JSON format easy to get at the end of the command, but for sure won't be from stdout.

Bottom line, I'll change the code to generate a real LOG file in JSON or YAML file, and this module should read the JSON/YAML into dictionary by sample and values, what should make easier the parsing.

@oneillkza, sorry to change that at these stage. I'll ping here when I update that and hopefully you can get it done when you come back!

Thanks both for all the help!

**Comment by ewels** [comment]
Ok, sounds great! Don't feel you have to do all of this stuff, it's certainly not a requirement :) But it does make downstream processing simpler..

I'll hold on this PR until you say it's ready to go then.

**Comment by oneillkza** [comment]
@lpantano No worries! As Phil has said, the parsing itself is literally one line of code, though it will need a bit of tweaking. Those changes sound worthwhile doing, and I'm glad we could get a prototype up and running to kick off this discussion.

**Comment by lpantano** [comment]
@ewels , thanks, I am working on this, I think you have great points.

I decided to use JSON, there will be information for each file/sample, but I was wondering how I should add the command line and version there? Commented lines?

The structure will be something like:

```
{"sample1":
 "metric1": value,
....
}
```

any pointers on how to add that information would be helpful !

thanks!

**Comment by ewels** [comment]
My suggestion would be just to move the JSON stuff up a level onto its own key. Then you can add in whatever else you want (and it's more future proofed):

```json
{
  "meta": {
    "tool": "mirTop",
    "version": "v0.1.3",
    "homepage": "https://github.com/miRTop/mirtop/"
  },
  "metrics": {
    "sample1": {
      "metric1": 12,
      "metric2": 34
    }
  }
}
```

This doesn't really affect MultiQC at all, we just parse the entire JSON object and then access the dictionary key for the sample data.

**Comment by lpantano** [comment]
@oneillkza , I updated the test file in the multiqc_testdata. So you are ready to go now! :)

**Comment by lpantano** [comment]
@oneillkza, I added the new examples to ewels/MultiQC_TestData, so you are all set to make the changes. Ping me with whatever you'll need! thanks a lot!

**Comment by ewels** [comment]
Hi all, any progress on this PR?

**Comment by oneillkza** [comment]
@ewels Sorry! I changed jobs in October, and this has been dangling at the back of my to-do list every since!

I'm having a go at it right now.

**Comment by ewels** [comment]
aha - great! And congratulations on the new job :) Let me know if you need any help.. Looks like your fork has fallen behind a bit, so probably good to pull upstream changes first..

**Comment by oneillkza** [comment]
Thanks!

And lol, just fixed the updating part.

The tiny issue I'm having now is that it looks like Lorena changed the mirtop output format (eg [in the test data](https://github.com/ewels/MultiQC_TestData/tree/master/data/modules/mirtop)) from CSV to JSON.

I wrote the original code based off another module that reads in CSV; I'll just need to figure out how to read in JSON instead. Glancing at another module (the kat module) this actually looks like it'll make things easier, since I don't need to manually write parsing code any more.

**Comment by ewels** [comment]
Absolutely: JSON and YAML files are the best, as they can be loaded into Python dicts in a single line. You then just have to think about reshaping them as necessary to be the format that MultiQC needs for plots and tables 👍

**Comment by lpantano** [comment]
Thank @oneillkza!, can you send a link or screenshot with the output? 

I remember to have a comment to separate the numbers in three figures: (*counts, *sum, *mean). And change then the plot to a barplot. Do you think would be that possible? 

I don't remember to have a decision about the name, @ewels is ok at is it now?

Congrats in the new job! And thanks for doing this!

**Comment by oneillkza** [comment]
@lpantano it's a bit big to easily make screenshots, but I've emailed an html example to your Harvard email address.

You can also generate it by checking out oneillkza/MultiQC and running MultiQC over the example data (in ewels/MultiQC_TestData)


Per my comments above, I definitely think we should separate it into three figures. It might also be good to highlight the stats for the reference miRNAs in a separate colour to the isomiRs. This is next on my todo list.

I don't think a barplot will work for cases where there are hundreds of samples (e.g. a TCGA cohort). 

Re: file naming, right now the test data contains two files named `mirtop_stats2.log` and   `mirtop_stats.log`. Right now, the search pattern is simply "does any file contain the word 'mirtop'". If there is expected to be a standard file name (e.g. mirtop_stats.log), or even if it's expected that the files will always have the extension .log, that would be much more efficient in practice.

**Comment by ewels** [comment]
As @oneillkza says - file name patterns are the fastest and easiest for MultiQC (as long as they're specific enough). For example, `*_fastqc.zip` is ideal.

File contents are fine though, especially if we know that the string we're looking for is within the first few lines of content. We still have the overhead of opening files and looking at the contents, but at least we can break this pattern if it's not found within the first `n` lines, without looking through the entire file. A header line comment is ideal. But not sure if you can have comment lines in JSON?

File: multiqc/modules/mirtop/mirtop.py (hunk 7 above)
Comment (oneillkza) [reply multiqc/modules/mirtop/mirtop.py]: Resolved in 5e3cee23daaff1105961dee4b9ae786d21d04f99

**Comment by lpantano** [comment]
hey @oneillkza!

**Comment by lpantano** [comment]
hey @oneillkza!

do you have any updates on this? I could help in 3 weeks or so if you need it. We will have time, since @ewels will be a daddy soon (i think) so it is not urgent, but it would be good to know the plan so we can have it done at some point :)

**Comment by ewels** [comment]
Hi all,

Any updates here? I'm back from leave now and will be putting out a MultiQC release soon.

Phil

**Comment by lpantano** [comment]
Hi, ok. @FlorianThibord is working on this, but depending on when you are doing the release it may be too soon, he is defending his thesis soon. Do you have an idea on when will it be? If not, for sure, will be done by the next one, now that Florian is working on this.

**Comment by ewels** [comment]
I'm only working 50% and have other projects as well, so it could take me a while yet. I'm hoping that it'll only be a week or two though. My plan is to do one release after the PRs are merged and any major bug issues are resolved, and then probably another not too much later once I've had time to go through some of the other recent issues and feature requests.

**Comment by FlorianThibord** [comment]
Hi @ewels ,
I took over the mirtop module project, and I think it's working nicely (no errors in --lint mode): https://github.com/FlorianThibord/MultiQC 
I worked on the latest branch of mutliqc, instead of the branch from @oneillkza to be sure that there won't be conflicts with the new version. But I acknowledged @oneillkza 's contribution in the changelog. I hope it's ok?  Should I open a new PR?

**Comment by oneillkza** [comment]
@FlorianThibord no objections to that from my side. Thanks for taking this over!

**Comment by ewels** [comment]
Hi @florianduclot,

Sorry for the slow reply. Yes please - open a new PR and I'll close this one (assuming that they are trying to achieve the same thing) 👍 

Phil

**Comment by florianduclot** [comment]
Hi @ewels,
I believe you meant to tag @FlorianThibord so doing it here.

**Comment by ewels** [comment]
Oops, sorry! Being too quick with the GitHub autocomplete 😬

**Comment by FlorianThibord** [comment]
Hi @ewels,
Thanks! I will open the new PR today or tomorrow.

Florian
